pkg/admission/prep_downscale.go

```diff
@@ -155,27 +164,54 @@ func prepareDownscale(ctx context.Context, logger log.Logger, ar v1.AdmissionRev
 			)
 		}
 
-		foundSts, err := findDownscalesDoneMinTimeAgo(stsList, ar.Request.Name)
-		if err != nil {
-			level.Warn(logger).Log("msg", "downscale not allowed due to error while parsing downscale annotations", "err", err)
-			return deny(
-				"downscale of %s/%s in %s from %d to %d replicas is not allowed because parsing downscale annotations failed.",
-				ar.Request.Resource.Resource, ar.Request.Name, ar.Request.Namespace, *oldReplicas, *newReplicas,
-			)
-		}
-		if foundSts != nil {
-			msg := fmt.Sprintf("downscale of %s/%s in %s from %d to %d replicas is not allowed because statefulset %v was downscaled at %v and is labelled to wait %s between zone downscales",
-				ar.Request.Resource.Resource, ar.Request.Name, ar.Request.Namespace, *oldReplicas, *newReplicas, foundSts.name, foundSts.lastDownscaleTime, foundSts.waitTime)
-			level.Warn(logger).Log("msg", msg, "err", err)
-			return deny(msg)
-		}
+		// If using zoneTracker instead of downscale annotations, check the zone file for recent downscaling.
+		// Otherwise, check the downscale annotations.
+		if zt != nil {
+			if err := zt.loadZones(ctx, stsList); err != nil {
+				level.Warn(logger).Log("msg", "downscale not allowed due to error while loading zones", "err", err)
+				return deny(
+					"downscale of %s/%s in %s from %d to %d replicas is not allowed because loading zones failed.",
+					ar.Request.Resource.Resource, ar.Request.Name, ar.Request.Namespace, *oldReplicas, *newReplicas,
+				)
+			}
+			// Check if the zone has been downscaled recently.
+			foundSts, err := zt.findDownscalesDoneMinTimeAgo(stsList, ar.Request.Name)
+			if err != nil {
+				level.Warn(logger).Log("msg", "downscale not allowed due to error while parsing downscale timestamps from the zone file", "err", err)
+				return deny(
+					"downscale of %s/%s in %s from %d to %d replicas is not allowed because parsing parsing downscale timestamps from the zone file failed.",
```

Comment:
do we need a log line here and the returning of a warning? couldn't it be just returning the result, or an empty result and an error? then we only deal with logging once where ever it is we're  calling `prepareDownscale` if the result is that we're denying the downscale

Reply:
I think that's fair, generally I'm just trying to follow the pattern we've implemented elsewhere in this webhook. I don't think the extra logging has any real negative trade-offs as far as I can tell and this way depending on if you're looking at the operator directly or the admission responses you have visibility into the root cause of the failure.